### PR TITLE
[SPARK-48233][SS][TESTS] Tests for streaming on columns with non-default collations

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

This change covers tests for streaming operations under columns of string type that are collated with non-utf8-binary collations. PR introduces following tests:
1) Non-stateful streaming for non-binary collated columns. We use `UTF8_BINARY_LCASE` non-binary collation as the input and assert that streaming propagates collation and that filtering behaves under rules of given collation.
2) Stateful streaming for binary collations. We use `UNICODE` collation as source and make sure that stateful operations (deduplication as taken as the example) work.
3) More tests that assert that stateful operations in combination with non-binary collations throw proper exception.

### Why are the changes needed?

You can find more information about collation effort in document attached to root jira ticket.

This PR adds tests for basic non-stateful streaming operations with collations (e.g. filtering).

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

PR is test only.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala [hunk 1]
```diff
@@ -484,6 +486,52 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
...
+      // scalastyle:on
+      assertNumStateRows(total = 2, updated = 1, droppedByWatermark = 0)
+    )
+  }
+
+  test("non-binary collation aware deduplication not supported") {
+    val inputData = MemoryStream[(String)]
+    val result = inputData.toDF()
+      .select(col("value")
+        .try_cast(StringType("UTF8_BINARY_LCASE")).as("str"))
```
Comment (jose-torres) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:517]: I'm a bit confused by this - is the test name flipped or is UTF8_BINARY_LCASE considered a non-binary collation?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 2]
```diff
@@ -1364,6 +1364,35 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging wi
...
+          |) USING parquet""".stripMargin)
+
+      val streamDf = spark.readStream.table("parquet_streaming_tbl")
+      val filteredDf = streamDf.filter("key = 'aaa'")
+
+      val clock = new StreamManualClock()
+      testStream(filteredDf)(
+        StartStream(triggerClock = clock, trigger = Trigger.ProcessingTime(100)),
+        Execute { _ =>
+          spark.createDataFrame(Seq("aaa" -> 1, "AAA" -> 2, "bbb" -> 3, "aa" -> 4))
```
Comment (jose-torres) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:1384]: I think it's important to also test a scenario where the incoming stream has a non-default collation itself.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:517]: AFAIK, LCASE means comparing as lowercase, so yes it's bound to non-binary equality.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala [hunk 3]
```diff
@@ -484,6 +486,52 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
       CheckLastBatch(("c", 9, "c"))
     )
   }
+
+  test("collation aware deduplication") {
+    val inputData = MemoryStream[(String, Int)]
+    val result = inputData.toDF()
+      .select(col("_1")
+        .try_cast(StringType("UNICODE")).as("str"),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:494]: While we are here, I see UNICODE is binary equality but non-binary ordering. Does this still ensure that we can put this into RocksDB which key is binary sorted and find the key group based on prefix of key including this column?

E.g. Say we have two columns, dept (String with UNICODE collation), session start (timestamp) as grouping key, and want to scan all grouping keys which are having dept as 'dept1'. This is required for several operations like session window aggregation.

My gut feeling is yes, but I would like to double confirm.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:1384]: Not sure I understand. streamDf has non-default collation, UTF8_BINARY_LCASE. What do you mean by "incoming" stream?

**Review by HeartSaVioR — COMMENTED** [review]
Let's make clear the scope of tests we are adding here. I see the PR title is about "stateless" but you are also aware that deduplication is "stateful". While I agree that we probably won't want to add the collation test for all stateful operators, let's make the scope more clear in PR title.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 3 above)
Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:494]: In theory, it should work, but we need to test this as well.
binary ordering means that you can use binary representation to check which string alphabetically comes first. But if we care only about equality (which is usually used in "groupings" and joins), binary equality is only important.

I will follow up with additional testing.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 1 above)
Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:517]: That's right, UTF8_BINARY_LCASE is a non-binary collation.

In other words, "AAA" and "aaa" are considered equal, even though binary representations are clearly different.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 2 above)
Comment (dbatomic) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:1384]: +1 to what @HeartSaVioR said. Idea was exactly to use `UTF8_BINARY_LCASE` in the source.

**Comment by dbatomic** [comment]
> Let's make clear the scope of tests we are adding here. I see the PR title is about "stateless" but you are also aware that deduplication is "stateful". While I agree that we probably won't want to add the collation test for all stateful operators, let's make the scope more clear in PR title.



> Let's make clear the scope of tests we are adding here. I see the PR title is about "stateless" but you are also aware that deduplication is "stateful". While I agree that we probably won't want to add the collation test for all stateful operators, let's make the scope more clear in PR title.



> Let's make clear the scope of tests we are adding here. I see the PR title is about "stateless" but you are also aware that deduplication is "stateful". While I agree that we probably won't want to add the collation test for all stateful operators, let's make the scope more clear in PR title.

Right, I updated both PR title and PR description. And yes, tests for collations are still pretty ad-hoc/selective.

Goal of this PR is to assert that basics work. As we create more thorough plan for collations and streaming we will start adding better organized test strategies.

Let me know if you think now is a good time to start with this. I was also thinking about creating new test suite only for collations, but that seemed like an overkill for this change.

**Review by jose-torres — APPROVED** [review]
Looks good from my side

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala (hunk 2 above)
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala:1384]: Makes sense, I think I misread the test case the first time.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:494]: Thanks, using UNICODE in grouping key of session window aggregation should cover it.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**